### ISSUE TITLE
deadline has elapsed against QA Canary Environment

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#24780]
### What version of Materialize are you using?

v0.84

### What is the issue?

In https://buildkite.com/materialize/nightlies/builds/6193#018d4857-9cae-4717-9619-831de6be3f18  the QA Canary Environment Base Load test failed during a simple SELECT:
```
2024-01-27 00:41:22 UTC	> SELECT COUNT(DISTINCT count_star) FROM qa_canary_environment.public_loadgen.sales_product_product_category WHERE count_distinct_product_id < 0
2024-01-27 00:41:22 UTC	
2024-01-27 00:41:22 UTC	
2024-01-27 00:41:22 UTC	==> mzcompose: test case workflow-default failed: <class 'materialize.ui.CommandFailureCausedUIError'>: running docker compose failed (exit status 1)
2024-01-27 00:41:22 UTC	16:1: error: deadline has elapsed
2024-01-27 00:41:22 UTC	     |
2024-01-27 00:41:22 UTC	  15 |
2024-01-27 00:41:22 UTC	  16 | > SELECT COUNT(DISTINCT count_star) FROM qa_canary_environment.public_loadgen.sales_product_product_category WHERE count_distinct_product_id < 0
2024-01-27 00:41:22 UTC	     | ^
```
This is running against the actual Mz cloud, so I'm taking such flakes pretty seriously: https://console.materialize.com/regions/aws-us-east-1/clusters/u146/qa_canary_environment_storage?timePeriod=1440 This is with Testdrive using a default timeout of 120 seconds.

**Comment by mjibson** [comment]
2m does seem like we should be able to handle it? Is this maybe a load generator that has been running for many months, but is on a very small cluster and so is genuinely CPU limited and can't process this in time?

**Comment by def-** [comment]
The storage cluster is https://console.materialize.com/regions/aws-us-east-1/clusters/u146/qa_canary_environment_storage?timePeriod=1440, which doesn't seem to be too overloaded to me, but sometimes does spike to 100% CPU usage for a bit.

**Comment by chaas** [comment]
Cool stuff - I was able to track down the query in the activity log and then in the statement lifecycle. Interesting results:

| statement_id | event_type | occurred_at |
| ------------- | ------------- |  ------------- |
| aaf71b69-5039-4ec6-99fe-a4f690e31fc3 | execution-began | 2024-01-27 00:39:22.254+00 |
| aaf71b69-5039-4ec6-99fe-a4f690e31fc3 | execution-finished | 2024-01-27 00:41:22.387+00 |
| aaf71b69-5039-4ec6-99fe-a4f690e31fc3 | compute-dependencies-finished | 2024-01-27 00:48:14.964+00 |
| aaf71b69-5039-4ec6-99fe-a4f690e31fc3 |storage-dependencies-finished | 2024-01-27 00:48:14.95+00 |

My interpretation:
* Execution finished b/c the query was canceled, but I guess it didn't actually spin down the dataflows so we still logged when the compute & storage dependencies were ready (9 mins later) after the fact.
* At 00:28 (see screenshot 1 & 2 below) both the `qa_canary_environment_storage` and `default` clusters were at almost 100 CPU and had "not ready" events then restarted.
    * **Do replicas automatically restart when CPU gets pegged? Or is something else causing the restart?**
* We can't look at workflow graphs back in time (if we really needed I could manually run the underlying queries), but I'm guessing after the restart, everything had to rehydrate so the query was blocked for 10mins during rehydration.

TLDR: CPU is maxing out. Should we decrease the load or increase the CPU of this replica?

I looked at the last 30 day graph (screenshot 3) and this same pattern of CPU maxing followed by a restart has been happening at this same time intermittently presumably from the test through that entire period.

I'm also seeing that the underlying source for this query is currently stalled FYI https://console.materialize.com/regions/aws-us-east-1/sources/u3872/qa_canary_environment/public_loadgen/product/workflow. Looking at `mz_internal.mz_source_status_history`, looks like it started happening yesterday at `2024-01-31 20:42:42` with `BrokerTransportFailure` so probably unrelated.

<img width="1000" alt="Screenshot 2024-02-01 at 4 01 46 PM" src="https://github.com/MaterializeInc/materialize/assets/4186354/d549bb68-8b78-4331-b25b-157247c2f85b">
<img width="1013" alt="Screenshot 2024-02-01 at 4 03 09 PM" src="https://github.com/MaterializeInc/materialize/assets/4186354/d9ed8cb6-4c5b-45b7-8181-435c79ebd4ea">
<img width="1032" alt="Screenshot 2024-02-01 at 4 02 56 PM" src="https://github.com/MaterializeInc/materialize/assets/4186354/92bcd7c5-5915-42f9-ac25-8abb087e0079">

**Comment by chaas** [comment]
Actually I looked into this even more, and these graphs are a bit misleading. Cross referencing grafana and slack, it looks like the restarts are happening first, due to canary & prod deployments or OOMs unrelated to this test, then the CPU and memory spikes are caused by rehydration after the restart.

This test ran right after a canary deployment https://materializeinc.slack.com/archives/CTESPM7FU/p1706315484935969
From grafana, the restart kicked off at 19:29 EST and CPU spiked up and didn't drop back down until 19:49 EST, the same time at which the cluster's memory plateaued (hydration finished). This test ran at 2024-01-26 19:41 EST, so during the rehydration period.
Seems like the root cause was actually just unfortunate timing with a deployment.

**Comment by chaas** [comment]
Closing since it looks like a one-off issue based on timing with a deployment. Feel free to reopen if this is continuing to occur
